mempool/src/pool/tx_verifier/chainstate_handle.rs

```diff
@@ -0,0 +1,243 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::BTreeMap;
+
+use chainstate::{
+    chainstate_interface::ChainstateInterface,
+    tx_verifier::{
+        transaction_verifier::storage::HasTxIndexDisabledError, TransactionVerifierStorageRef,
+    },
+    ChainstateError,
+};
+use chainstate_types::storage_result;
+use common::{
+    chain::{
+        tokens::{TokenAuxiliaryData, TokenId},
+        Block, DelegationId, GenBlock, OutPoint, OutPointSourceId, PoolId, Transaction,
+        TxMainChainIndex,
+    },
+    primitives::{Amount, Id},
+};
+use pos_accounting::{AccountingBlockUndo, DelegationData, PoSAccountingView, PoolData};
+use subsystem::blocking::BlockingHandle;
+use utils::shallow_clone::ShallowClone;
+use utxo::{Utxo, UtxosBlockUndo, UtxosStorageRead, UtxosView};
+
+/// Chainstate handle error includes errors coming from chainstate and inter-subprocess
+/// communication errors.
+#[derive(thiserror::Error, PartialEq, Eq, Debug)]
+pub enum Error {
+    #[error("Chainstate error: {0}")]
+    Chainstate(#[from] ChainstateError),
+    #[error("Subsystem error: {0}")]
+    Subsystem(#[from] subsystem::subsystem::CallError),
+}
+
+impl HasTxIndexDisabledError for Error {
+    fn tx_index_disabled_error() -> Self {
+        Self::Chainstate(ChainstateError::tx_index_disabled_error())
+    }
+}
+
+impl From<pos_accounting::Error> for Error {
+    fn from(e: pos_accounting::Error) -> Self {
+        Error::from(ChainstateError::from(chainstate::BlockError::from(e)))
+    }
+}
+
+impl From<utxo::Error> for Error {
+    fn from(e: utxo::Error) -> Self {
+        chainstate::tx_verifier::TransactionVerifierStorageError::from(e).into()
+    }
+}
+
+impl From<chainstate::TransactionVerifierStorageError> for Error {
+    fn from(e: chainstate::TransactionVerifierStorageError) -> Self {
+        Error::from(ChainstateError::from(chainstate::BlockError::from(e)))
+    }
+}
+
+impl From<Error> for crate::error::Error {
+    fn from(e: Error) -> Self {
+        match e {
+            Error::Chainstate(e) => crate::error::Error::TxValidationError(e.into()),
+            Error::Subsystem(_) => crate::error::Error::SubsystemFailure,
+        }
+    }
+}
+
+/// A wrapper over handle to chanistate
+pub struct ChainstateHandle<C>(BlockingHandle<C>);
+
+impl<C> ChainstateHandle<C> {
+    pub fn new(inner: subsystem::Handle<C>) -> Self {
+        Self(BlockingHandle::new(inner))
+    }
+}
+
+impl<C: 'static + Send> ChainstateHandle<C> {
+    /// Chainstate subsystem call. We assume chainstate is alive here.
+    pub fn call<R: 'static + Send>(
+        &self,
+        func: impl 'static + Send + FnOnce(&C) -> Result<R, ChainstateError>,
```

Comment:
The challenge is that some of these statics are contravariant... which is OK. So trying to make everything the same lifetime will be too constrained and it won't work.

Reply:
This is a chainstate handle which is currently only instantiated with `Box<dyn ChainstateInterface>`. There may be merit in simplifying the code by just using the `Box<dyn ChainstateInterface>` type directly without generics. We don't lose the ability to mock since `dyn ChainstateInterface` is still there. Thoughts?

Reply:
Well, any Deref<ChainstateInterface> implements ChainstateInterface... I don't think mocks will be affected. So with my wild imagination I'm thining we could generalize this further and use generic lifetimes instead of 'static, because with 'static only that Box will work, so there's a question to the benefit of this.

It sounds worse to just use Box<dyn ChainstateInterface> directly, but when we put everything in context, it will look like that we should either generalize lifetimes or just use that Box directly, unless there are plans in the future to change something relevant and having this kind of generic type with non-generic lifetime helps with that.

I don't know what the right answer is... but it doesn't seem right as it's now.

Reply:
As far as this `call` method is concerned, I only use it to implement the `tx_verifier` traits. It is sufficient for that use case. I made it private so it's clear we don't need it to be as general as possible, just general enough to deal with stuff that uses it locally.

On the broader question of `ChainstateHandle`, that too I consider an implementation detail. It lives in a private submodule of `mempool::tx_verifier` so it is not accessible from outside and mempool only accesses it through the transaction verifier traits. So at least for now, it is merely an implementation detail. Do you envision it being more generally useful?

Reply:
That's fair. It's just that the intent from the design is unclear. I'm OK with it... not a big deal.